src/compiler/crystal/macros/methods.cr

```diff
@@ -798,6 +798,17 @@ module Crystal
         interpret_argless_method(method, args) { StringLiteral.new(@value.underscore) }
       when "upcase"
         interpret_argless_method(method, args) { StringLiteral.new(@value.upcase) }
+      when "parse_type_name"
+        interpret_argless_method(method, args) do
+          type_name = if @value.includes? '|'
+                        # Rewrite `Type1|Type2` to `Union(Type1, Type2)`
+                        %(Union(#{@value.gsub '|', ','}))
```

Comment:
Such a simple rewrite is insufficient. `Array(Foo | Bar)` would become an invalid `Union(Array(Foo, Bar))`.

We need to employ `parse_bare_proc_type` which performs exactly what's necessary here.

Reply:
Ahh, I had a feeling it wasn't that simple.  I pushed up a commit that uses that method and adds a spec for this context.